From 744eb29abe15fb126613b1711e1870b46a252682 Mon Sep 17 00:00:00 2001
From: Kevin Brace <kevinbrace@gmx.com>
Date: Wed, 6 Jul 2016 22:02:17 -0500
Subject: [PATCH 06/21] Limiting IGA2 Address Mode Selection bit access to
 UniChrome Pro or later

This bit is located at 3X5.62[0]. Linear mode is always used. It is
very likely that UniChrome (CLE266, KM400, P4M800, etc.) does not
implement this bit.

Signed-off-by: Kevin Brace <kevinbrace@gmx.com>
---
 src/via_display.c | 10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/src/via_display.c b/src/via_display.c
index 385daed..9c2acc3 100644
--- a/src/via_display.c
+++ b/src/via_display.c
@@ -2157,8 +2157,14 @@ viaIGA2SetDisplayRegister(ScrnInfoPtr pScrn, DisplayModePtr mode)
     /* Use IGA2 for DVP1 Data Source Selection 0. */
     ViaCrtcMask(hwp, 0x9B, 0x10, 0x10);
 
-    /* Linear Mode */
-    ViaCrtcMask(hwp, 0x62, 0x00, 0x01);
+    /* UniChrome Pro or later */
+    if ((pVia->Chipset != VIA_CLE266) && (pVia->Chipset != VIA_KM400)) {
+        /* Set IGA2 to linear mode. */
+        /* 3X5.62[0] - IGA2 Address Mode Selection
+         *             0: Linear
+         *             1: Tile */
+        ViaCrtcMask(hwp, 0x62, 0x00, 0x01);
+    }
 
 
     /* Set IGA2 horizontal total pixels.*/
-- 
1.9.1

